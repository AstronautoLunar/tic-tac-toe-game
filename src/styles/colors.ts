const colors = {
  playerX: '#46A6FF',
  playerO: '#FF465C',
  detach: '#6246FF',
  background: '#060029',
  secundary: '#E8E5F9',
  success: '#51EE4E'
};

export default colors;