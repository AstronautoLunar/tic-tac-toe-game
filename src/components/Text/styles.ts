import styled from 'styled-components/native';

import colors from '../../styles/colors';

export type SpanProps = {
  fontSize?: string | number;
  color?: string;
  margin?: string;
}

export const Span = styled.Text<SpanProps>`
  font-size: ${props => props.fontSize ? props.fontSize : 24}px;
  font-weight: bold;
  
  text-align: center;

  margin: ${props => props.margin ? props.margin : '0 24px'};

  color: ${props => props.color ? props.color : colors.secundary};
`;