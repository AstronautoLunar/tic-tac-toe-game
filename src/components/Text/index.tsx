import { Span, SpanProps } from "./styles";

type AditionalProps = {
  children: string;
}

type TextProps = AditionalProps & SpanProps

function Text({ 
  children, 
  fontSize, 
  color 
}: TextProps) {
  return (
    <Span fontSize={fontSize} color={color}>
      { children }
    </Span>
  )
}

export default Text;