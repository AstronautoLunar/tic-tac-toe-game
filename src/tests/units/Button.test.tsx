import styled from 'styled-components/native';
import { render } from '@testing-library/react-native';

import colors from '../../styles/colors';

import Text, { SpanProps } from '../../components/Text';

export type ButtonAreaProps = {
  backgroundColor?: string;
  border?: string;
  marginTop?: string | number;
  marginBottom?: string | number;
  marginLeft?: string | number;
  marginRight?: string | number;
  width?: string;
  height?: string;
  flex?: number;
};

type OptionsButton = {
  colorButton?: ButtonAreaProps['backgroundColor'];
  BorderButton?: ButtonAreaProps['border'];
  colorText?: SpanProps['color'];
  sizeText?: SpanProps['fontSize'];
  marginTopButton?: ButtonAreaProps['marginTop'];
  marginBottomButton?: ButtonAreaProps['marginBottom'];
  marginLeftButton?: ButtonAreaProps['marginLeft'];
  marginRightButton?: ButtonAreaProps['marginRight'];
  marginTotalText?: SpanProps['margin'];
  widthButton?: ButtonAreaProps['width'];
  heightButton?: ButtonAreaProps['height'];
  flexButton?: ButtonAreaProps['flex'];
};

function createConfigButtonLength(prop: 'width' | 'height') {
  return (props: ButtonAreaProps) => {
    if (props[prop]) {
      return props[prop];
    } else if (props.flex) {
      return 'auto'
    } else {
      return '100%'
    }
  }
}

const ButtonArea = styled.View<ButtonAreaProps>`
  width: ${createConfigButtonLength('width')};
  height: ${createConfigButtonLength('height')};
  flex: ${props => props.flex ? props.flex : 'none'};

  display: flex;
  justify-content: center;
  align-items: center;

  background-color: ${props => props.backgroundColor ? props.backgroundColor : colors.detach};

  margin-top: ${props => props.marginTop ? props.marginTop : 0}px;
  margin-bottom: ${props => props.marginBottom ? props.marginBottom : 0}px;
  margin-left: ${props => props.marginLeft ? props.marginLeft : 0}px;
  margin-right: ${props => props.marginRight ? props.marginRight : 0}px;

  border: ${props => props.border ? props.border : '0px'};
  border-radius: 10px;
`;



type ButtonProps = {
  children: string | string[];
  optionsButton?: OptionsButton;
}

function Button({ children, optionsButton }: ButtonProps) {
  return (
    <ButtonArea
      backgroundColor={optionsButton?.colorButton}
      border={optionsButton?.BorderButton}
      marginBottom={optionsButton?.marginBottomButton}
      marginTop={optionsButton?.marginTopButton}
      marginLeft={optionsButton?.marginLeftButton}
      marginRight={optionsButton?.marginRightButton}
      width={optionsButton?.widthButton}
      height={optionsButton?.heightButton}
      flex={optionsButton?.flexButton}
    >
      <Text
        fontSize={optionsButton?.sizeText}
        color={optionsButton?.colorText}
        margin={optionsButton?.marginTotalText}
      >
        {children}
      </Text>
    </ButtonArea>
  )
}

describe('Test Unit in button component', () => {

  test('When applying setting for flex on button, remove width and height automatically', () => {
    // SET
    const CONFIGURATION_FLEX_BUTTON = {
      flexButton: 1
    };

    // ACT
    const result = render(
      <Button
        optionsButton={CONFIGURATION_FLEX_BUTTON}
      >
        Clique me
      </Button>
    ).toJSON();

    // ASSERT
    expect(result).toMatchSnapshot();
  });

  test('When applying configuration for width and height on button, remove flex configuration automatically', () => {
    // SET
    const CONFIGURATION_WIDTH_HEIGHT_BUTTON = {
      widthButton: '70%',
      heightButton: '70px'
    }

    // ACT
    const result = render(
      <Button
        optionsButton={CONFIGURATION_WIDTH_HEIGHT_BUTTON}
      >
        Clique me
      </Button>
    ).toJSON();

    // ASSERT
    expect(result).toMatchSnapshot();
  });
});