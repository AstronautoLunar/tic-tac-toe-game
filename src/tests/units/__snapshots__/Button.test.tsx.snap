// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Unit in button component When applying configuration for width and height on button, remove flex configuration automatically 1`] = `
<View
  height="70px"
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#6246FF",
        "borderColor": "black",
        "borderRadius": 10,
        "borderStyle": "solid",
        "borderWidth": 0,
        "display": "flex",
        "flexBasis": "auto",
        "flexGrow": 0,
        "flexShrink": 0,
        "height": 70,
        "justifyContent": "center",
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
        "width": "70%",
      },
    ]
  }
  width="70%"
>
  <Text
    style={
      Array [
        Object {
          "color": "#E8E5F9",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginBottom": 0,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 0,
          "textAlign": "center",
        },
      ]
    }
  >
    Clique me
  </Text>
</View>
`;

exports[`Test Unit in button component When applying setting for flex on button, remove width and height automatically 1`] = `
<View
  flex={1}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#6246FF",
        "borderColor": "black",
        "borderRadius": 10,
        "borderStyle": "solid",
        "borderWidth": 0,
        "display": "flex",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "height": "auto",
        "justifyContent": "center",
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
        "width": "auto",
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#E8E5F9",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginBottom": 0,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 0,
          "textAlign": "center",
        },
      ]
    }
  >
    Clique me
  </Text>
</View>
`;
